### DISCUSSION TITLE
Add FFZ Emote Effects

### DISCUSSION THREAD

**Discussion by Nerixyz** [Chatterino/chatterino2#4429] [Ideas] [answered]
[FFZ added effects for emotes](https://www.frankerfacez.com/subscribe). If Chatterino supports other platform's subscription features (BTTV/7TV personal emotes), it should support this one as well.

[emote-effects.webm](https://user-images.githubusercontent.com/19953266/222928182-05b62cd5-8e2e-4a2a-b25b-098ebf3cf567.webm)

---

The effects seem to be special emotes returned in https://api.frankerfacez.com/v1/set/global (with `modifier` and `modifier_flags`). https://github.com/FrankerFaceZ/FrankerFaceZ/commit/e433aa3340b3d8e975542cc50656e2f3c5d932dd implemented this in the FFZ extension.

**Comment by jupjohn** [comment]
This feels equivalent (ideal implementation-wise) to how c7 does namepaints. My only issue here is that FFZ expects the client to know how to modify the emotes instead of sending the modification CSS back in the API's response (IIRC 7TV does that). I wouldn't be a fan of hardcoding support for these in chatterino and having to release new versions just to add new modifier support.

That being said, I doubt the FFZ team are going to want to make such a change now that they've implemented it this way.

**Comment by atrigent** [comment]
I just created #4443 which would also be needed to make use of these specific emotes (since they are user-limited supporter emotes).

Another couple things to note: in the FFZ API docs (https://api.frankerfacez.com/docs/) it states:

> I realize that implementing these effects can potentially be a lot of work for a developer, especially if they aren't using a client that renders chat via HTML and CSS. In the future I'd like to potentially offer a server endpoint that prerenders emotes with effects. Unfortunately, we lack the resources to do that at this time.

Since they've considered going so far as prerendering the effects, it doesn't seem like they'd be necessarily opposed to providing an API that just provides a bit of CSS code. So I think it'd be worth asking for.

It also says:

> For now, please at least check for the Hidden flag to avoid rendering modifiers that should be hidden. I think that would give users without full support in their client of choice the best experience.

So this could be a good place to start.

**Reply by Mm2PL** [reply]
> API that just provides a bit of CSS code

I don't think we can even use CSS without implementing a web browser into our layout/message rendering pipeline. We basically use a mostly custom widget renderer

**Reply by LosFarmosCTL** [reply]
> I don't think we can even use CSS without implementing a web browser into our layout/message rendering pipeline.

We definitely can't, I had the fun of making a custom CSS gradient implementation when doing paints for c7, but some standardized API response about the effect would remove the need for hardcoding the effects, even if it still requires us to manually parse + render it.

**Comment by Felanbird** [comment, marked as answer]
> If Chatterino supports other platform's subscription features (BTTV/7TV personal emotes), it should support this one as well.

We don't support these features.

See a more detailed answer in https://github.com/Chatterino/chatterino2/discussions/4444#discussioncomment-5267936
